sidecar/src/agent_remote_config.rs

```diff
@@ -0,0 +1,212 @@
+            handle: Mutex::new(Some(handle.clone().map()?)),
+        },
+        handle,
+    ))
+}
+
+pub fn new_reader(endpoint: &Endpoint) -> AgentRemoteConfigReader<NamedShmHandle> {
+    AgentRemoteConfigReader {
+        handle: open_named_shm(endpoint).ok(),
+        endpoint: Some(endpoint.clone()),
+        current_config: None,
+    }
+}
+
+pub fn reader_from_shm(handle: ShmHandle) -> io::Result<AgentRemoteConfigReader<ShmHandle>> {
+    Ok(AgentRemoteConfigReader {
+        handle: Some(handle.map()?),
+        endpoint: None,
+        current_config: None,
+    })
+}
+
+pub fn new_writer(endpoint: &Endpoint) -> io::Result<AgentRemoteConfigWriter<NamedShmHandle>> {
+    Ok(AgentRemoteConfigWriter {
+        handle: Mutex::new(Some(
+            NamedShmHandle::create(path_for_endpoint(endpoint), 0x1000)?.map()?,
+        )),
+    })
+}
+
+pub trait ReaderOpener<T>
+where
+    T: FileBackedHandle<T>,
+{
+    fn open(endpoint: &Endpoint) -> Option<MappedMem<T>>;
+}
+
+fn open_named_shm(endpoint: &Endpoint) -> io::Result<MappedMem<NamedShmHandle>> {
+    NamedShmHandle::open(path_for_endpoint(endpoint))?.map()
+}
+
+impl ReaderOpener<NamedShmHandle> for AgentRemoteConfigReader<NamedShmHandle> {
+    fn open(endpoint: &Endpoint) -> Option<MappedMem<NamedShmHandle>> {
+        open_named_shm(endpoint).ok()
+    }
+}
+
+impl ReaderOpener<ShmHandle> for AgentRemoteConfigReader<ShmHandle> {
+    fn open(_: &Endpoint) -> Option<MappedMem<ShmHandle>> {
+        None
+    }
+}
+
+impl<T: FileBackedHandle<T> + From<MappedMem<T>>> AgentRemoteConfigReader<T>
+where
+    AgentRemoteConfigReader<T>: ReaderOpener<T>,
+{
+    // bool is true when it changed
+    pub fn read<'a>(&'a mut self) -> (bool, &[u8]) {
+        if let Some(ref handle) = self.handle {
+            let source_data: &RawData = handle.as_slice().into();
+            let new_generation = source_data.meta.generation;
+
+            let fetch_data = |reader: &'a mut AgentRemoteConfigReader<T>| {
+                let size = std::mem::size_of::<RawMetaData>() + source_data.meta.size;
+
+                let handle = reader.handle.take().unwrap().ensure_space(size);
+                reader.handle.replace(handle);
+                let handle = reader.handle.as_ref().unwrap();
+
+                let mut new_mem = Vec::with_capacity(size);
+                new_mem.extend_from_slice(&handle.as_slice()[0..size]);
+
+                // refetch, might have been resized
+                let new_data: &RawData = handle.as_slice().into();
+                let copied_data: &RawData = new_mem.as_slice().into();
+
+                // Ensure the next write hasn't started yet *and* the data is from the expected generation
+                if !new_data.meta.writing.load(Ordering::SeqCst)
+                    && new_generation == copied_data.meta.generation
+                {
+                    reader.current_config.replace(new_mem);
+                    return Some((true, new_data.as_slice()));
```

Comment:
Here you return the slice pointing to the memory mapped region, but on line 171 you rerturn the copied Vec.

To prevent reads concurrent with writes, I think you should return the copied_data buffer instead